logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java

```diff
@@ -0,0 +1,248 @@
+package org.logstash.common.io;
+
+import org.logstash.ackedqueue.Queueable;
+import org.logstash.ackedqueue.SequencedList;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.zip.CRC32;
+import java.util.zip.Checksum;
+
+// TODO: currently assuming continuous seqNum is the byte buffer where we can deduct the maxSeqNum from the min + count.
+// TODO: we could change this and support non-continuous seqNums but I am not sure we should.
+// TODO: checksum is not currently computed.
+
+public class ByteBufferPageIO implements PageIO {
+    public static final byte VERSION = 1;
+    public static final int CHECKSUM_SIZE = Integer.BYTES;
+    public static final int LENGTH_SIZE = Integer.BYTES;
+    public static final int SEQNUM_SIZE = Long.BYTES;
+    public static final int MIN_RECORD_SIZE = SEQNUM_SIZE + LENGTH_SIZE + CHECKSUM_SIZE;
+    public static final int HEADER_SIZE = 1;     // version byte
+    static final List<byte[]> EMPTY_READ = new ArrayList<>(0);
+
+    private final int capacity;
+    private final List<Integer> offsetMap; // has to be extendable
+    private final ByteBuffer buffer;
+    private long minSeqNum; // TODO: to make minSeqNum final we have to pass in the minSeqNum in the constructor and not set it on first write
```

Comment:
yeah, I would defer that refactor - I don't think it is urgent we do that now...